functions/src/etl/FirestoreAuditor.ts

```diff
@@ -0,0 +1,86 @@
+import equal from 'deep-equal';
+import { firestore } from 'firebase-admin';
+import * as functions from 'firebase-functions';
+import { Change, EventContext } from 'firebase-functions';
+import { FirestoreAdmin } from '../../../shared/src/firebase/FirestoreAdmin';
+import DocumentSnapshot = firestore.DocumentSnapshot;
+
+/**
+ * Watches write updates in collection and subcollections and inserts the previous record value into a
+ * "{collectionName}-history" collection. In there, for each document id we maintain a history subcollection
+ * with containing all the previous versions of a document.
+ */
+export class FirestoreAuditor {
+	readonly firestoreAdmin: FirestoreAdmin;
+
+	constructor(firestoreAdmin: FirestoreAdmin) {
+		this.firestoreAdmin = firestoreAdmin;
+	}
+
+	/**
+	 * Triggers changes to documents of root collections
+	 */
+	auditCollectionTrigger = functions.firestore
+		.document('{collectionId}/{docId}')
+		.onWrite(async (change: Change<DocumentSnapshot>, context: EventContext) => {
+			return this.auditFirestore(change, context);
+		});
+
+	/**
+	 * Triggers changes to documents of sub collections
+	 */
+	auditSubcollectionTrigger = functions.firestore
```

Comment:
Is this needed?

Couldn't we have a single trigger like this one:

```ts
auditCollectionTrigger = functions.firestore
		.document('{collectionId}/{document=**}')
		.onWrite(async (change: Change<DocumentSnapshot>, context: EventContext) => {
			return this.auditFirestore(change, context);
		});
```

This should imo work, see: https://firebase.google.com/docs/firestore/security/rules-structure#recursive_wildcards

Reply:
Then the case of an update in a nested subcollection would also be covered.

Reply:
Ah nice, that would be much better. Let me try that 👍

Reply:
Works like a charm 👍